// @vite/vitejs/es5/abi/coder/bytes.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FixedBytesCoder = exports.BytesCoder = exports.DynamicBytesCoder = void 0;
var index_1 = require("./index");
var utils_1 = require("../utils");
var DynamicBytesCoder = (function (_super) {
    __extends(DynamicBytesCoder, _super);
    function DynamicBytesCoder(type, localName) {
        return _super.call(this, type, type, localName, true) || this;
    }
    DynamicBytesCoder.prototype.defaultValue = function () {
        return '';
    };
    DynamicBytesCoder.prototype.encode = function (writer, value) {
        value = (0, utils_1.arrayify)(value);
        var length = writer.writeValue(value.length);
        length += writer.writeBytes(value);
        return length;
    };
    DynamicBytesCoder.prototype.decode = function (reader) {
        return reader.readBytes(reader.readValue().toNumber(), true);
    };
    return DynamicBytesCoder;
}(index_1.Coder));
exports.DynamicBytesCoder = DynamicBytesCoder;
var BytesCoder = (function (_super) {
    __extends(BytesCoder, _super);
    function BytesCoder(localName) {
        return _super.call(this, 'bytes', localName) || this;
    }
    BytesCoder.prototype.decode = function (reader) {
        return reader.coerce(this, (0, utils_1.hexlify)(_super.prototype.decode.call(this, reader)));
    };
    return BytesCoder;
}(DynamicBytesCoder));
exports.BytesCoder = BytesCoder;
var FixedBytesCoder = (function (_super) {
    __extends(FixedBytesCoder, _super);
    function FixedBytesCoder(size, localName) {
        var _this = this;
        var name = "bytes".concat(String(size));
        _this = _super.call(this, name, name, localName, false) || this;
        _this.size = size;
        return _this;
    }
    FixedBytesCoder.prototype.defaultValue = function () {
        return ('0000000000000000000000000000000000000000000000000000000000000000').substring(0, this.size * 2);
    };
    FixedBytesCoder.prototype.encode = function (writer, value) {
        var data = (0, utils_1.arrayify)(value);
        if (data.length !== this.size) {
            throw new Error("incorrect data length: ".concat(value));
        }
        return writer.writeBytes(data);
    };
    FixedBytesCoder.prototype.decode = function (reader) {
        return reader.coerce(this, (0, utils_1.hexlify)(reader.readBytes(this.size)));
    };
    return FixedBytesCoder;
}(index_1.Coder));
exports.FixedBytesCoder = FixedBytesCoder;
