// @vite/vitejs/es5/WS/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WS_RPC = void 0;
var ipc_ws_1 = require("../communication/ipc_ws");
var Websocket = require('websocket').w3cwebsocket;
var WsRpc = (function (_super) {
    __extends(WsRpc, _super);
    function WsRpc(path, timeout, options) {
        if (path === void 0) { path = 'ws://localhost:23457'; }
        if (timeout === void 0) { timeout = 60000; }
        if (options === void 0) { options = {
            protocol: '',
            headers: '',
            clientConfig: {
                keepalive: true,
                keepaliveInterval: 30 * 1000
            }
        }; }
        var _this = _super.call(this, {
            onEventTypes: ['error', 'close', 'connect'],
            sendFuncName: 'send',
            path: path
        }) || this;
        if (!path) {
            throw _this.ERRORS.CONNECT(path);
        }
        _this.timeout = timeout;
        _this.protocol = options.protocol;
        _this.headers = options.headers;
        _this.clientConfig = options.clientConfig;
        _this._destroyed = false;
        _this.reconnect();
        return _this;
    }
    WsRpc.prototype.reconnect = function () {
        var _this = this;
        if (this._destroyed)
            return;
        this.socket && (this.socket.onclose = function () { });
        this.disconnect();
        this.socket = new Websocket(this.path, this.protocol, null, this.headers, null, this.clientConfig);
        this.socket.onopen = function () {
            (_this.socket.readyState === _this.socket.OPEN) && _this._connected();
        };
        this.socket.onclose = function () {
            _this._closed();
        };
        this.socket.onerror = function (err) {
            _this._errored(err);
        };
        this.socket.onmessage = function (e) {
            var data = (typeof e.data === 'string') ? e.data : '';
            _this._parse([data]);
        };
    };
    WsRpc.prototype.disconnect = function () {
        this.socket && this.socket.close && this.socket.close();
        this.socket = null;
    };
    WsRpc.prototype.destroy = function () {
        this.remove('error');
        this.remove('close');
        this.remove('connect');
        this.disconnect();
        this._destroyed = true;
    };
    return WsRpc;
}(ipc_ws_1.default));
exports.WS_RPC = WsRpc;
exports.default = exports.WS_RPC;
